### PR TITLE
[glsl-out]: combined image sampler

### PR DESCRIPTION
**Pull request by JCapucho**
I have also included some changes that are related to the fact that we were using some vulkan glsl only parts

### CODE REVIEW

File: src/back/glsl.rs [hunk 1]
```diff
@@ -61,23 +62,32 @@ pub struct Options {
     pub entry_point: (String, ShaderStage),
 }
 
+#[derive(Debug, Clone)]
+pub struct TextureMapping {
+    pub texture: Handle<GlobalVariable>,
+    pub sampler: Handle<GlobalVariable>,
```
Comment (kvark) [inline src/back/glsl.rs:68]: this should be `Option<>`, since not all textures are used with samplers

File: src/back/glsl.rs [hunk 2]
```diff
@@ -1632,3 +1663,191 @@ fn write_format_glsl(format: StorageFormat) -> &'static str {
...
+    functions: &FastHashMap<Handle<Function>, String>,
+) -> Result<Vec<TextureMapping>, Error> {
+    fn collect_texture_mapping_expr(
+        func: &Function,
+        expr: Handle<Expression>,
+        mappings: &mut Vec<TextureMapping>,
+    ) -> Result<(), Error> {
+        match func.expressions[expr] {
+            Expression::Access { base, index } => {
+                collect_texture_mapping_expr(func, base, mappings)?;
```
Comment (kvark) [inline src/back/glsl.rs:1678]: We have the logic of walking through expressions in `proc::Interface`. It would be nice to share this code wider by separating the visitor logic from the traversal.

File: src/back/glsl.rs [hunk 3]
```diff
@@ -1632,3 +1663,191 @@ fn write_format_glsl(format: StorageFormat) -> &'static str {
...
+                    crate::SampleLevel::Exact(expr) | crate::SampleLevel::Bias(expr) => {
+                        collect_texture_mapping_expr(func, expr, mappings)?
+                    }
+                    crate::SampleLevel::Auto => {}
+                }
+                if let Some(expr) = depth_ref {
+                    collect_texture_mapping_expr(func, expr, mappings)?;
+                }
+
+                let mapping = mappings.iter().find(|map| map.texture == tex_handle);
```
Comment (kvark) [inline src/back/glsl.rs:1728]: If I understand correctly, this is an internal map we are collecting. It looks like we are often searching it by the texture handle. Would it make things easier if this internal map was `FastHashMap<Handle<GlobalVar>, Handle<GlobalVar>>`? (addressable by texture handles)

File: src/back/glsl.rs [hunk 4]
```diff
@@ -1596,6 +1620,13 @@ fn builtin_to_glsl(builtin: BuiltIn) -> &'static str {
     }
 }
 
+fn get_base_ty<'a>(types: &'a Arena<Type>, ty: &'a TypeInner) -> &'a TypeInner {
+    match ty {
+        TypeInner::Pointer { base, .. } => get_base_ty(types, &types[*base].inner),
```
Comment (kvark) [inline src/back/glsl.rs:1625]: this function specifically works with pointers? why did you need it?

File: src/back/glsl.rs [hunk 5]
```diff
@@ -214,13 +223,145 @@ pub fn write<'a>(module: &'a Module, out: &mut impl Write, options: Options) ->
 
     writeln!(out)?;
 
-    let mut globals_lookup = FastHashMap::default();
+    let mut functions = FastHashMap::default();
 
-    for ((handle, global), usage) in module.global_variables.iter().zip(func.global_usage.iter()) {
-        if usage.is_empty() {
+    for (handle, func) in module.functions.iter() {
```
[the remark this answers is not in the document]
Comment (JCapucho) [reply src/back/glsl.rs:228]: Right now this exposes all the functions except for the ones that are entry points

File: src/back/glsl.rs (hunk 4 above)
Comment (JCapucho) [reply src/back/glsl.rs:1629]: I had some issues where i got pointers where other types where expected this is a temporary solution

File: src/back/glsl.rs [hunk 6]
```diff
@@ -1,10 +1,12 @@
 use super::{BorrowType, MaybeOwned};
 use crate::{
+    proc::{Interface, Visitor},
     Arena, ArraySize, BinaryOperator, BuiltIn, Constant, ConstantInner, DerivativeAxis, Expression,
     FastHashMap, Function, FunctionOrigin, GlobalVariable, Handle, ImageClass, Interpolation,
     IntrinsicFunction, LocalVariable, MemberOrigin, Module, ScalarKind, ShaderStage, Statement,
     StorageAccess, StorageClass, StorageFormat, StructMember, Type, TypeInner, UnaryOperator,
 };
+use log::warn;
```
Comment (kvark) [inline src/back/glsl.rs:9]: note that in other places we just use the full path for logging, i.e. `log::warn!("...")`

File: src/back/glsl.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (kvark) [reply src/back/glsl.rs:229]: Filed #186 to follow-up

File: src/back/glsl.rs (hunk 4 above)
Comment (kvark) [reply src/back/glsl.rs:1621]: Please tell more. I thought we fixed this in #100. If not, this should be fixed in the front-ends, not worked around here.

File: src/back/glsl.rs [hunk 7]
```diff
@@ -1624,3 +1659,100 @@ fn write_format_glsl(format: StorageFormat) -> &'static str {
...
+                    Expression::GlobalVariable(global) => global,
+                    // Temp
+                    Expression::Load { pointer } => match self.expressions[pointer] {
+                        Expression::GlobalVariable(global) => global,
+                        _ => unreachable!(),
+                    },
+                    _ => unreachable!(),
+                };
+
+                let mapping = self.map.get(&tex_handle);
```
Comment (kvark) [inline src/back/glsl.rs:1693]: let's use the entry API. This would be `match self.map.entry(tex_handle) {...}`

File: src/back/glsl.rs [hunk 8]
```diff
@@ -1624,3 +1659,100 @@ fn write_format_glsl(format: StorageFormat) -> &'static str {
...
+
+                if mapping.map_or(false, |sampler| *sampler != None) {
+                    self.error = Some(Error::Custom(String::from(
+                        "Cannot use texture with two different samplers",
+                    )));
+                }
+
+                if mapping.is_none() {
```
Comment (kvark) [inline src/back/glsl.rs:1724]: same here

**Review by kvark — CHANGES_REQUESTED** [review]
Looks great! We just need one more iteration on it.

File: src/back/glsl.rs (hunk 4 above)
Comment (JCapucho) [reply src/back/glsl.rs:1621]: I just checked right now `test-data/spv/cube.frag.spv` and the frontend is producing multiple loads to globals

File: src/back/glsl.rs [hunk 9]
```diff
@@ -743,7 +802,14 @@ fn write_expression<'a, 'b>(
             module.borrow_type(builder.locals[handle].ty),
         ),
         Expression::Load { pointer } => {
-            write_expression(&builder.expressions[pointer], module, builder)?
+            let (expr, ty) = write_expression(&builder.expressions[pointer], module, builder)?;
+
+            let ty = match *ty.borrow() {
+                TypeInner::Pointer { base, .. } => base,
```
Comment (kvark) [inline src/back/glsl.rs:808]: I think what we want is to not have pointers where they are implied. I.e. when accessing any global or local variables, doing pointer expression indirection is pointless

File: src/back/glsl.rs [hunk 10]
```diff
@@ -1344,46 +1383,54 @@ fn write_type<'a>(
...
+    {
+        return Err(Error::Custom(String::from(
+            "Arrayed non 2d images aren't supported",
+        )));
+    }
+
+    Ok(format!(
+        "{}{}{}{}{}",
+        match kind {
+            ScalarKind::Sint => "i",
```
Comment (kvark) [inline src/back/glsl.rs:1410]: let's use the `map_scalar` function here

**Review by kvark — APPROVED** [review]
Looks great! I have a few minor-ish notes left

**Review by kvark — APPROVED** [review]
